client.go

```diff
@@ -180,26 +157,53 @@ func DefaultBackoff(min, max time.Duration, attemptNum int, resp *http.Response)
 	return sleep
 }
 
+// nopCloser is like ioutil.NopCloser except it preserves Seek
+type nopCloser struct {
+	io.ReadSeeker
+}
+
+func (nopCloser) Close() error { return nil }
+
+func (c nopCloser) Seek(offset int64, whence int) (int64, error) {
+	return c.ReadSeeker.Seek(offset, whence)
+}
+
 // Do wraps calling an HTTP method with retries.
-func (c *Client) Do(req *Request) (*http.Response, error) {
+func (c *Client) Do(req *http.Request) (*http.Response, error) {
 	c.Logger.Printf("[DEBUG] %s %s", req.Method, req.URL)
 
+	if req.Body != nil {
+		// if body is not rewindable then read it to a buffer that is rewindable
+		if readSeeker, ok := req.Body.(io.ReadSeeker); ok {
+			// always override any underlying Close function on req.Body so that
+			// net/http does not close it after the first try
```

Comment:
If this is preventing net/http from closing after the first attempt, doesn't it also mean that the body will now never be closed? I would think you'd want to defer a close on the underlying body before wrapping.